roslibrust/src/ros1/subscriber.rs

```diff
@@ -0,0 +1,219 @@
+use super::tcpros::ConnectionHeader;
+use abort_on_drop::ChildTask;
+use roslibrust_codegen::RosMessageType;
+use std::{marker::PhantomData, sync::Arc};
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::TcpStream,
+    sync::{broadcast, RwLock},
+};
+
+pub struct Subscriber<T> {
+    receiver: broadcast::Receiver<Vec<u8>>,
+    phantom: PhantomData<T>,
+}
+
+impl<T: RosMessageType> Subscriber<T> {
+    pub(crate) fn new(receiver: broadcast::Receiver<Vec<u8>>) -> Self {
+        Self {
+            receiver,
+            phantom: PhantomData,
+        }
+    }
+
+    pub async fn next(&mut self) -> Result<T, Box<dyn std::error::Error>> {
+        let data = self.receiver.recv().await.map_err(|err| Box::new(err))?;
+        Ok(serde_rosmsg::from_slice(&data[..]).map_err(|err| Box::new(err))?)
```

Comment:
So if I'm reading this right (which I may not be). If we have multiple subscribers to a topic, they each get a copy of the raw over the wire data and separately deserialize the data. That is totally fine for now, but would prefer if we deserde the data once and the held typed channels? I guess the complication there is that on the send side of the channel we need to know the type or create a type erasure?

Totally a valid impl for this stage, but maybe I'll file an issue to clean this up eventually.

Reply:
Right, the type erasure here is a little complicated and I decided to punt that problem since it's an optimization that's separate from setting up the subscriber.